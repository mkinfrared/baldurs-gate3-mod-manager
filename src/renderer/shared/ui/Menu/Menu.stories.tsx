/* eslint-disable react-hooks/rules-of-hooks */
import { Meta, StoryObj } from "@storybook/react";
import { createRef, useRef, useState } from "react";

import Button from "renderer/shared/ui/Button";
import MenuItem from "renderer/shared/ui/MenuItem";

import Menu from "./Menu";
import css from "./Story.module.scss";

type Story = StoryObj<typeof meta>;

const meta = {
  title: "UI/Menu",
  component: Menu,
  tags: ["autodocs"],
} satisfies Meta<typeof Menu>;

const Default: Story = {
  args: {
    id: "marklar",
    comboboxRef: createRef(),
  },
  render: ({ id }) => {
    const [showMenu, setShowMenu] = useState(false);
    const ref = useRef<HTMLButtonElement>(null);

    const toggleMenu = () => {
      setShowMenu((prevState) => !prevState);
    };

    return (
      <div className={css.Story}>
        <Button buttonRef={ref} onClick={toggleMenu}>
          Show
        </Button>
        {showMenu && (
          <Menu id={id} comboboxRef={ref} onClose={() => setShowMenu(false)}>
            <MenuItem value={1}>Foobar-Lorem-ipsum-dolor-set-amet-1</MenuItem>
            <MenuItem value={2}>Foobar-Lorem-ipsum-dolor-set-amet-2</MenuItem>
            <MenuItem value={3}>Foobar-Lorem-ipsum-dolor-set-amet-3</MenuItem>
            <MenuItem value={4}>Foobar-Lorem-ipsum-dolor-set-amet-4</MenuItem>
            <MenuItem value={5}>Foobar-Lorem-ipsum-dolor-set-amet-5</MenuItem>
            <MenuItem value={6}>Foobar-Lorem-ipsum-dolor-set-amet-6</MenuItem>
            <MenuItem value={7}>Foobar-Lorem-ipsum-dolor-set-amet-7</MenuItem>
            <MenuItem value={8}>Foobar-Lorem-ipsum-dolor-set-amet-8</MenuItem>
            <MenuItem value={9}>Foobar-Lorem-ipsum-dolor-set-amet-9</MenuItem>
            <MenuItem value={10}>Foobar-Lorem-ipsum-dolor-set-amet-10</MenuItem>
            <MenuItem value={11}>Foobar-Lorem-ipsum-dolor-set-amet-11</MenuItem>
            <MenuItem value={12}>Foobar-Lorem-ipsum-dolor-set-amet-12</MenuItem>
            <MenuItem value={13}>Foobar-Lorem-ipsum-dolor-set-amet-13</MenuItem>
            <MenuItem value={14}>Foobar-Lorem-ipsum-dolor-set-amet-14</MenuItem>
            <MenuItem value={15}>Foobar-Lorem-ipsum-dolor-set-amet-15</MenuItem>
            <MenuItem value={16}>Foobar-Lorem-ipsum-dolor-set-amet-16</MenuItem>
            <MenuItem value={17}>Foobar-Lorem-ipsum-dolor-set-amet-17</MenuItem>
            <MenuItem value={18}>Foobar-Lorem-ipsum-dolor-set-amet-18</MenuItem>
            <MenuItem value={19}>Foobar-Lorem-ipsum-dolor-set-amet-19</MenuItem>
            <MenuItem value={20}>Foobar-Lorem-ipsum-dolor-set-amet-20</MenuItem>
            <MenuItem value={21}>Foobar-Lorem-ipsum-dolor-set-amet-21</MenuItem>
            <MenuItem value={22}>Foobar-Lorem-ipsum-dolor-set-amet-22</MenuItem>
            <MenuItem value={23}>Foobar-Lorem-ipsum-dolor-set-amet-23</MenuItem>
            <MenuItem value={24}>Foobar-Lorem-ipsum-dolor-set-amet-24</MenuItem>
            <MenuItem value={25}>Foobar-Lorem-ipsum-dolor-set-amet-25</MenuItem>
            <MenuItem value={26}>Foobar-Lorem-ipsum-dolor-set-amet-26</MenuItem>
            <MenuItem value={27}>Foobar-Lorem-ipsum-dolor-set-amet-27</MenuItem>
            <MenuItem value={28}>Foobar-Lorem-ipsum-dolor-set-amet-28</MenuItem>
            <MenuItem value={29}>Foobar-Lorem-ipsum-dolor-set-amet-29</MenuItem>
            <MenuItem value={30}>Foobar-Lorem-ipsum-dolor-set-amet-30</MenuItem>
            <MenuItem value={31}>Foobar-Lorem-ipsum-dolor-set-amet-31</MenuItem>
            <MenuItem value={32}>Foobar-Lorem-ipsum-dolor-set-amet-32</MenuItem>
            <MenuItem value={33}>Foobar-Lorem-ipsum-dolor-set-amet-33</MenuItem>
            <MenuItem value={34}>Foobar-Lorem-ipsum-dolor-set-amet-34</MenuItem>
            <MenuItem value={35}>Foobar-Lorem-ipsum-dolor-set-amet-35</MenuItem>
            <MenuItem value={36}>Foobar-Lorem-ipsum-dolor-set-amet-36</MenuItem>
            <MenuItem value={37}>Foobar-Lorem-ipsum-dolor-set-amet-37</MenuItem>
            <MenuItem value={38}>Foobar-Lorem-ipsum-dolor-set-amet-38</MenuItem>
            <MenuItem value={39}>Foobar-Lorem-ipsum-dolor-set-amet-39</MenuItem>
            <MenuItem value={40}>Foobar-Lorem-ipsum-dolor-set-amet-40</MenuItem>
            <MenuItem value={41}>Foobar-Lorem-ipsum-dolor-set-amet-41</MenuItem>
            <MenuItem value={42}>Foobar-Lorem-ipsum-dolor-set-amet-42</MenuItem>
            <MenuItem value={43}>Foobar-Lorem-ipsum-dolor-set-amet-43</MenuItem>
            <MenuItem value={44}>Foobar-Lorem-ipsum-dolor-set-amet-44</MenuItem>
            <MenuItem value={45}>Foobar-Lorem-ipsum-dolor-set-amet-45</MenuItem>
            <MenuItem value={46}>Foobar-Lorem-ipsum-dolor-set-amet-46</MenuItem>
            <MenuItem value={47}>Foobar-Lorem-ipsum-dolor-set-amet-47</MenuItem>
            <MenuItem value={48}>Foobar-Lorem-ipsum-dolor-set-amet-48</MenuItem>
            <MenuItem value={49}>Foobar-Lorem-ipsum-dolor-set-amet-49</MenuItem>
            <MenuItem value={50}>Foobar-Lorem-ipsum-dolor-set-amet-50</MenuItem>
            <MenuItem value={51}>Foobar-Lorem-ipsum-dolor-set-amet-51</MenuItem>
            <MenuItem value={52}>Foobar-Lorem-ipsum-dolor-set-amet-52</MenuItem>
            <MenuItem value={53}>Foobar-Lorem-ipsum-dolor-set-amet-53</MenuItem>
            <MenuItem value={54}>Foobar-Lorem-ipsum-dolor-set-amet-54</MenuItem>
            <MenuItem value={55}>Foobar-Lorem-ipsum-dolor-set-amet-55</MenuItem>
            <MenuItem value={56}>Foobar-Lorem-ipsum-dolor-set-amet-56</MenuItem>
            <MenuItem value={57}>Foobar-Lorem-ipsum-dolor-set-amet-57</MenuItem>
            <MenuItem value={58}>Foobar-Lorem-ipsum-dolor-set-amet-58</MenuItem>
            <MenuItem value={59}>Foobar-Lorem-ipsum-dolor-set-amet-59</MenuItem>
            <MenuItem value={60}>Foobar-Lorem-ipsum-dolor-set-amet-60</MenuItem>
            <MenuItem value={61}>Foobar-Lorem-ipsum-dolor-set-amet-61</MenuItem>
            <MenuItem value={62}>Foobar-Lorem-ipsum-dolor-set-amet-62</MenuItem>
            <MenuItem value={63}>Foobar-Lorem-ipsum-dolor-set-amet-63</MenuItem>
            <MenuItem value={64}>Foobar-Lorem-ipsum-dolor-set-amet-64</MenuItem>
            <MenuItem value={65}>Foobar-Lorem-ipsum-dolor-set-amet-65</MenuItem>
            <MenuItem value={66}>Foobar-Lorem-ipsum-dolor-set-amet-66</MenuItem>
            <MenuItem value={67}>Foobar-Lorem-ipsum-dolor-set-amet-67</MenuItem>
            <MenuItem value={68}>Foobar-Lorem-ipsum-dolor-set-amet-68</MenuItem>
            <MenuItem value={69}>Foobar-Lorem-ipsum-dolor-set-amet-69</MenuItem>
            <MenuItem value={70}>Foobar-Lorem-ipsum-dolor-set-amet-70</MenuItem>
            <MenuItem value={71}>Foobar-Lorem-ipsum-dolor-set-amet-71</MenuItem>
            <MenuItem value={72}>Foobar-Lorem-ipsum-dolor-set-amet-72</MenuItem>
            <MenuItem value={73}>Foobar-Lorem-ipsum-dolor-set-amet-73</MenuItem>
            <MenuItem value={74}>Foobar-Lorem-ipsum-dolor-set-amet-74</MenuItem>
            <MenuItem value={75}>Foobar-Lorem-ipsum-dolor-set-amet-75</MenuItem>
            <MenuItem value={76}>Foobar-Lorem-ipsum-dolor-set-amet-76</MenuItem>
            <MenuItem value={77}>Foobar-Lorem-ipsum-dolor-set-amet-77</MenuItem>
            <MenuItem value={78}>Foobar-Lorem-ipsum-dolor-set-amet-78</MenuItem>
            <MenuItem value={79}>Foobar-Lorem-ipsum-dolor-set-amet-79</MenuItem>
            <MenuItem value={80}>Foobar-Lorem-ipsum-dolor-set-amet-80</MenuItem>
            <MenuItem value={81}>Foobar-Lorem-ipsum-dolor-set-amet-81</MenuItem>
            <MenuItem value={82}>Foobar-Lorem-ipsum-dolor-set-amet-82</MenuItem>
            <MenuItem value={83}>Foobar-Lorem-ipsum-dolor-set-amet-83</MenuItem>
            <MenuItem value={84}>Foobar-Lorem-ipsum-dolor-set-amet-84</MenuItem>
            <MenuItem value={85}>Foobar-Lorem-ipsum-dolor-set-amet-85</MenuItem>
            <MenuItem value={86}>Foobar-Lorem-ipsum-dolor-set-amet-86</MenuItem>
            <MenuItem value={87}>Foobar-Lorem-ipsum-dolor-set-amet-87</MenuItem>
            <MenuItem value={88}>Foobar-Lorem-ipsum-dolor-set-amet-88</MenuItem>
            <MenuItem value={89}>Foobar-Lorem-ipsum-dolor-set-amet-89</MenuItem>
            <MenuItem value={90}>Foobar-Lorem-ipsum-dolor-set-amet-90</MenuItem>
            <MenuItem value={91}>Foobar-Lorem-ipsum-dolor-set-amet-91</MenuItem>
            <MenuItem value={92}>Foobar-Lorem-ipsum-dolor-set-amet-92</MenuItem>
            <MenuItem value={93}>Foobar-Lorem-ipsum-dolor-set-amet-93</MenuItem>
            <MenuItem value={94}>Foobar-Lorem-ipsum-dolor-set-amet-94</MenuItem>
            <MenuItem value={95}>Foobar-Lorem-ipsum-dolor-set-amet-95</MenuItem>
            <MenuItem value={96}>Foobar-Lorem-ipsum-dolor-set-amet-96</MenuItem>
            <MenuItem value={97}>Foobar-Lorem-ipsum-dolor-set-amet-97</MenuItem>
            <MenuItem value={98}>Foobar-Lorem-ipsum-dolor-set-amet-98</MenuItem>
            <MenuItem value={99}>Foobar-Lorem-ipsum-dolor-set-amet-99</MenuItem>
            <MenuItem value={100}>
              Foobar-Lorem-ipsum-dolor-set-amet-100
            </MenuItem>
          </Menu>
        )}
      </div>
    );
  },
};

export { Default };

export default meta;
