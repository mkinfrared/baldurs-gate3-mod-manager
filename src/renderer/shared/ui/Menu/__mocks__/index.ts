export { default } from "./Menu.mock";
