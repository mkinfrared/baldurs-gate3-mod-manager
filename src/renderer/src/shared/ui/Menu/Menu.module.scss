.Menu {
  --left: 0;
  --top: 0;
  --combobox-width: 0;
  --background: var(--surface-1);
  --border-radius: 0.4rem;
  --border-color: var(--primary-4);

  margin: 0;

  &[role="listbox"] {
    background-color: var(--background);
    border: 0.1rem solid var(--border-color);
    border-radius: var(--border-radius);
    box-shadow: 0 32px 64px -12px rgb(85 105 135 / 8%);
    left: var(--left);
    max-height: 25rem;
    max-width: 100%;
    min-height: 4.4rem;
    min-width: var(--combobox-width);
    overflow-y: auto;
    padding: 0;
    position: fixed;
    scroll-behavior: smooth;
    scroll-padding-block: 0.4rem;
    scroll-snap-type: y;
    top: var(--top);
    width: var(--combobox-width);
  }
}
