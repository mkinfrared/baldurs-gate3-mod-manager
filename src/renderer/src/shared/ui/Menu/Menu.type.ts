import { ReactNode, RefObject } from "react";

export type MenuProps = {
  children?: ReactNode;
  /**
   * a string that will be applied as a css class to parent element
   */
  className?: string;
  /**
   * a ref to an element that combobox will attach to during render
   */
  comboboxRef: RefObject<HTMLElement>;
  /**
   * a unique id to use in conjunction with aria-controls
   */
  id: string;
  /**
   * a callback which fires on outside click of Menu
   * to close the current Modal component
   */
  onClose: () => void;
};
