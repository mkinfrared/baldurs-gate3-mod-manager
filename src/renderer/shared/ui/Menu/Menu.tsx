import { useLayoutEffect, useRef } from "react";
import { createPortal } from "react-dom";

import { classNames } from "renderer/shared/lib/helpers";
import { useOutsideClick } from "renderer/shared/lib/hooks";

import css from "./Menu.module.scss";
import { MenuProps } from "./Menu.type";

// TODO fix Y position in iOS on scroll
const Menu = ({ className, children, comboboxRef, id, onClose }: MenuProps) => {
  const { body } = document;
  const ref = useRef<HTMLUListElement>(null);

  useOutsideClick([ref, comboboxRef], onClose);

  const computeVisualYAxis = (
    combobox: HTMLElement,
    dropdownMenu: HTMLElement,
  ) => {
    const { top, bottom } = combobox.getBoundingClientRect();
    const { height: comboboxHeight } = dropdownMenu.getBoundingClientRect();
    const nextBottomY = bottom + comboboxHeight;
    const nextTopY = top - comboboxHeight;

    // if dropdown bottom edge doesn't show up outside the viewport
    // place it at the bottom of combobox
    if (nextBottomY < window.visualViewport!.height) {
      dropdownMenu.style.setProperty("--top", `${bottom}px`);

      // else if dropdown top edge doesn't show up outside the viewport
      // place it at the top of combobox
    } else if (nextTopY >= 0) {
      dropdownMenu.style.setProperty("--top", `${top - comboboxHeight}px`);
      // otherwise place it at the bottom of combobox
    } else {
      dropdownMenu.style.setProperty("--top", `${bottom}px`);
    }
  };

  const computeYAxis = (combobox: HTMLElement, dropdownMenu: HTMLElement) => {
    if (window.visualViewport) {
      computeVisualYAxis(combobox, dropdownMenu);

      return;
    }

    const { top, bottom } = combobox.getBoundingClientRect();
    const { height: comboboxHeight } = dropdownMenu.getBoundingClientRect();
    const nextBottomY = bottom + comboboxHeight;
    const nextTopY = top - comboboxHeight;

    // if dropdown bottom edge doesn't show up outside the viewport
    // place it at the bottom of combobox
    if (nextBottomY < window.innerHeight) {
      dropdownMenu.style.setProperty("--top", `${bottom}px`);

      // else if dropdown top edge doesn't show up outside the viewport
      // place it at the top of combobox
    } else if (nextTopY >= 0) {
      dropdownMenu.style.setProperty("--top", `${top - comboboxHeight}px`);
      // otherwise place it at the bottom of combobox
    } else {
      dropdownMenu.style.setProperty("--top", `${bottom}px`);
    }
  };

  const computeXAxis = (combobox: HTMLElement, dropdownMenu: HTMLElement) => {
    const { left, right, width } = combobox.getBoundingClientRect();
    // const { width: comboboxWidth } = dropdownMenu.getBoundingClientRect();
    const nextRightX = left + width;
    const nextLeftX = right - width;

    // if dropdown right edge doesn't show up outside the viewport
    // place it at the left of combobox
    if (nextRightX < window.innerWidth) {
      dropdownMenu.style.setProperty("--left", `${left}px`);

      // else if dropdown left edge doesn't show up outside the viewport
      // place it at the right of combobox
    } else if (nextLeftX >= 0) {
      dropdownMenu.style.setProperty("--left", `${nextLeftX}px`);
      // otherwise place it at the left of viewport and set the width of viewport
    } else {
      dropdownMenu.style.setProperty("--left", "0px");
    }

    dropdownMenu.style.setProperty("--combobox-width", `${width}px`);
  };

  const setPosition = (comboboxElement: HTMLElement) => {
    const { current } = ref;

    if (!current) {
      return;
    }

    const { width } = comboboxElement.getBoundingClientRect();

    current.style.setProperty("--combobox-width", `${width}px`);

    computeYAxis(comboboxElement, current);

    computeXAxis(comboboxElement, current);
  };

  useLayoutEffect(() => {
    document.body.style.overflow = "hidden";

    if (comboboxRef.current) {
      setPosition(comboboxRef.current);
    }

    return () => {
      document.body.style.overflow = "";
    };
  }, [children, comboboxRef]);

  useLayoutEffect(() => {
    const handleResize = () => {
      if (comboboxRef.current) {
        setPosition(comboboxRef.current);
      }
    };

    window.addEventListener("resize", handleResize);

    window.visualViewport?.addEventListener("resize", handleResize);

    window.visualViewport?.addEventListener("scroll", () => {
      setTimeout(handleResize, 1000);
    });

    return () => {
      window.removeEventListener("resize", handleResize);

      window.visualViewport?.removeEventListener("resize", handleResize);

      window.visualViewport?.removeEventListener("scroll", handleResize);
    };
  }, []);

  return createPortal(
    <ul
      className={classNames(css.Menu, className)}
      data-testid="Menu"
      id={id}
      ref={ref}
      role="listbox"
      onKeyDown={() => null}
    >
      {children}
    </ul>,
    body,
  );
};

export default Menu;
