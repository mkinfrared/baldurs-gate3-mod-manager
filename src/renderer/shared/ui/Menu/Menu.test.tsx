import { render } from "@testing-library/react";

import Menu from "./Menu";

describe("<Menu />", () => {
  const comboboxRef = { current: null };
  const id = "marklar";
  const onClose = jest.fn();

  const Component = (
    <Menu comboboxRef={comboboxRef} id={id} onClose={onClose} />
  );

  it("should be defined", () => {
    expect(Menu).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should contain a data test id", () => {
    const { getByTestId } = render(Component);
    const element = getByTestId("Menu");

    expect(element).toBeDefined();
  });
});
