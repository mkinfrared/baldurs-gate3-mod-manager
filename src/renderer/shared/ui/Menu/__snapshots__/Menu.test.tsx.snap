// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> should match the snapshot 1`] = `<div />`;
